### PR TITLE
Fix for incorrect sample queries language displayed in GE

### PR DESCRIPTION
**Pull request by irvinesunday**
Fixes https://github.com/microsoftgraph/microsoft-graph-explorer-v4/issues/659

**Cause**: This happens when multiple requests are attempting to seed the sample queries simultaneously in the DevX API. This is an intermittent problem which is temporarily remedied by initiating a restart of the app service.

**Proposed fix**:
- Refactoring out the samples queries fetching process to a separate class, `SamplesStore.cs`; and injecting this class through DI.
- Using thread locking to ensure one locale cache is seeded by one executing thread at any given time.
- Adding a test utility library, `MockTestUtility`, for helping out with mocking of tests among test projects, and refactoring out common code into this library.
- Adding tests to validate the `SamplesStore.cs` class functionality.
- Substituting deprecated [functionality ](https://github.com/microsoftgraph/microsoft-graph-explorer-api/blob/89fa6533cf7dd89d7309381a492b82a7d9de20b3/GraphWebApi/Controllers/GraphExplorerSamplesController.cs#L316-L319)serving a deprecated method with a `NotImplementedException()`. This was taking a dependency on some of the fields that were deleted (due to the sample queries fetch functionality being refactored out to a separate class), hence throwing errors.

### CODE REVIEW

File: GraphExplorerSamplesService/Services/SamplesStore.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                // Localized copy of samples is to be seeded by only one executing thread.
+                lock (_samplesLock)
+                {
+                    /* Check whether a previous thread already seeded an
+                     * instance of the localized samples during the lock.
+                     */
+                    var lockedLocale = locale;
+                    var seededSampleQueriesList = _samplesCache.Get<SampleQueriesList>(lockedLocale);
+                    if (seededSampleQueriesList == null)
```
Comment (zengin) [inline GraphExplorerSamplesService/Services/SamplesStore.cs:58]: nit:
invert if with an early return to reduce indentation:
```
if (seededSampleQueriesList != null)
{
  return seededSampleQueriesList;
}

// lines 60-80 goes here with less indentation
```

File: GraphExplorerSamplesService/Services/SamplesStore.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+                        bool orderSamples = false;
+
+                        if (lockedLocale.Equals("en-us", StringComparison.OrdinalIgnoreCase))
+                        {
+                            /* Current business process only supports ordering of the English
+                               translation of the sample queries.
+                            */
+                            orderSamples = true;
+                        }
```
Comment (zengin) [inline GraphExplorerSamplesService/Services/SamplesStore.cs:77]:
```suggestion
                        /* Current business process only supports ordering of the English
                           translation of the sample queries.
                        */
                        bool orderSamples = lockedLocale.Equals("en-us", StringComparison.OrdinalIgnoreCase);
```

File: GraphWebApi/Controllers/GraphExplorerSamplesController.cs [hunk 3]
```diff
@@ -323,76 +308,14 @@ public async Task<IActionResult> DeleteSampleQueryAsync(string id)
...
-                // Save the JSON string to the source file
-                await _fileUtility.WriteToFile(policiesJson, _policiesFilePathSource);
-
-                // Return the list of policies
-                return policies;
-            }
-
-            // Return the list of policies
-            return SamplesPolicyService.DeserializeSampleQueriesPolicies(jsonFileContents);
+            throw new NotImplementedException();
```
Comment (zengin) [inline GraphWebApi/Controllers/GraphExplorerSamplesController.cs:318]: I don't see anything related to this in the description. Is this an intentional change?

Comment (irvinesunday) [reply GraphWebApi/Controllers/GraphExplorerSamplesController.cs:318]: This is a deprecated functionality serving a deprecated method. It was taking a dependency on some of the fields that were deleted (due to the sample queries fetch functionality being refactored out to a separate class), hence throwing errors. It didn't make sense to try and resolve these.

File: GraphExplorerSamplesService/Services/SamplesStore.cs (hunk 2 above)
Comment (irvinesunday) [reply GraphExplorerSamplesService/Services/SamplesStore.cs]: Sorry, your commit suggestion was rendered stale from the above commit. The suggested changes have been added though.

File: GraphExplorerSamplesService/Services/SamplesStore.cs (hunk 1 above)
Comment (irvinesunday) [reply GraphExplorerSamplesService/Services/SamplesStore.cs]: Resolved

**Comment by irvinesunday** [comment]
> > **Cause**: This happens when multiple requests are attempting to seed the sample queries simultaneously in the DevX API. This is an intermittent problem which is temporarily remedied by initiating a restart of the app service.
> 
> Are multiple requests coming from the same client, or are we mixing different sources of requests?
> 
> Do you mind providing a call pattern from client with corresponding actions on the service side and how singleton pattern is interrupting the same call pattern with the fix?

This is caused by multiple requests coming in from different clients with different `Accept-Language` headers. When one of the request is seeding the cache for its specific locale, the subsequent request coming in alters the locale variable state, hence the mix up in the cached name and the cached sample query data. Using a Singleton will enforce only a single access to the class for every request. This will help prevent inconsistent state by allowing us to perform thread locking.

**Pull request merged**